runtime/parachains/src/scheduler.rs

```diff
@@ -337,6 +320,12 @@ impl<T: Config> Module<T> {
 			}
 		});
 		ParathreadQueue::set(thread_queue);
+
+		// We want to re-schedule as though it is the first block of the session.
+		let now = <frame_system::Module<T>>::block_number() + One::one();
+		<SessionStartBlock<T>>::set(now);
```

Comment:
At this point I think we should alter the comment on `SessionStartBlock` to emphasize this subtlety, specifically that the session change is signalled at B and that B+1 is actually the first block to be executed in that session.